site/journal/2020/Understanding args in Glimmer Components.md

```diff
@@ -0,0 +1,309 @@
+
+```js
+root.haveABirthday();
+root.changeNameTo("C. D. Krycho");
+console.log(profile.description);
+// -> "C. D. Krycho Krycho is 34 years old"
+```
+
+But the “automatically” here is just normal JavaScript semantics: we passed in a reference to an object, so when we change the properties on that object and then invoke a getter which depends on those values, we get an updated value. While there are some differences in the specific details of the implementation (as I cover below in Part II), this is fundamentally how `args` works in Glimmer components. And notice: there’s no `@tracked` in sight, no template layer involved, and in fact nothing about Ember or Glimmer in sight. It is, as they say, *just JavaScript*™.
+
+This also means that we can see why certain things *won’t* work. I often see people write code like this both when writing new Glimmer components from scratch and when migrating from Ember components:
+
+```js
+class Profile extends Component {
+  constructor(args) {
+    super(args);
+    this.description =
+      `${this.args.name} is ${this.args.age} years old`;
+  }
+}
+```
+
+In this case, changes to `name` or `age` on `root` won’t be reflected in the `Profile` instance:
+
+```js
+let root = new Root();
+let profile = new Profile(root.user);
+console.log(profile.description);
+// -> "Chris Krycho is 33 years old"
+
+root.haveABirthday();
+root.changeNameTo("C. D. Krycho");
+console.log(profile.description);
+// -> "Chris Krycho is 33 years old"
+```
+
+Why not? Because assigning to an instance property in the `constructor` for `Profile` simply evaluates the values passed in and saves the result as a field, no different than writing `this.description = "hello"`. A getter, like `get description() { ... }` is re-executed when invoked as `profile.description`. A class field just has whatever value is has. Accordingly, changes to `root.user` cannot affect the value of `profile.description` in this approach.
+
+This brings us to the key takeaway for working with `args` in these components:
+
+<i>You should *never* assign from a property on `args` to a local class field in a Glimmer component—because changes to the parent will never be reflected in the component![^local-copy] Not because of anything special about Glimmer components or autotracking, but just because of how JavaScript works!</i>
+
+So far so good—but in a real Glimmer Component, we don’t pass in the args blob directly ourselves. Instead, we usually pass named argument values, like `@name={{this.user.name}}`. What’s more, we don’t necessarily pass reference types. We might pass in strings, numbers, etc. too—just as with `name`. So how does this still work?
+
+## Part II: The Real Implementation
+
+The answer is that the template layer wires this up for us—but in a way that, once again, we can understand in terms of plain old JavaScript. When you write a component invocation like this:
+
+```handlebars
+<Profile
+  @name={{this.name}}
```

Comment:
`{{this.user.name}}`?

Reply:
Haha, yeah, I've been refactoring this a couple different directions and missed a spot. Thank you!